src/Microsoft.Diagnostics.NETCore.Client/HandleableCollection.cs

```diff
@@ -0,0 +1,316 @@
+        /// </summary>
+        /// <remarks>
+        /// The returned enumerator is over a copy of the underlying collection so that there are no concurrency issues.
+        /// </remarks>
+        IEnumerator<T> IEnumerable<T>.GetEnumerator()
+        {
+            VerifyNotDisposed();
+
+            IList<T> copy;
+            lock (_items)
+            {
+                copy = _items.ToList();
+            }
+            return copy.GetEnumerator();
+        }
+
+        /// <summary>
+        /// Disposes the <see cref="HandleableCollection{T}"/> by clearing all items and handlers.
+        /// </summary>
+        /// <remarks>
+        /// All pending handlers with throw <see cref="ObjectDisposedException"/>.
+        /// </remarks>
+        public void Dispose()
+        {
+            if (!_disposed)
+            {
+                _disposalSource.Cancel();
+
+                lock (_items)
+                {
+                    _items.Clear();
+
+                    _handlers.Clear();
+                }
+
+                _disposalSource.Dispose();
+
+                _disposed = true;
+            }
+        }
+
+        /// <summary>
+        /// Adds an item so that it may be observed by a handler.
+        /// </summary>
+        /// <param name="item">Item to add to the collection.</param>
+        /// <remarks>
+        /// The item may be immediately consumed if a handler removes the item, thus it may
+        /// not be stored into the underlying list.
+        /// </remarks>
+        public void Add(in T item)
+        {
+            VerifyNotDisposed();
+
+            lock (_items)
+            {
+                bool handledValue = false;
+                for (int i = 0; !handledValue && i < _handlers.Count; i++)
+                {
+                    Tuple<TaskCompletionSource<T>, Handler> handler = _handlers[i];
+
+                    if (TryHandler(item, handler.Item2, handler.Item1, out handledValue))
+                    {
+                        _handlers.RemoveAt(i);
+                        i--;
+                    }
+                }
+
+                if (!handledValue)
+                {
+                    _items.Add(item);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns the first item offered to the handler
+        /// or waits for a future item if no item is immediately available.
+        /// </summary>
+        /// <param name="timeout">The amount of time to wait before cancelling the handler.</param>
+        /// <returns>The first item offered to the handler.</returns>
+        public T Handle(TimeSpan timeout)
+        {
+            VerifyNotDisposed();
+
+            return Handle(DefaultHandler, timeout);
+        }
+
+        /// <summary>
+        /// Returns the item on which the handler completes or waits for future items
+        /// if the handler does not immediately complete.
```

Comment:
`Task.Result` blocks until the Task is completed is some manner. It'll return with the result if ran to completion, or it will throw an exception if faulted or cancelled.